### PR TITLE
planner: Add Revision check in tryLockMDLAndUpdateSchemaIfNecessary

### LINKED ISSUE #54779: Significant CPU usage for tryLockMDLAndUpdateSchemaIfNecessary for insert statement on table over 1k columns
## Enhancement
Even with plan cache hits, high cpu on tryLockMDLAndUpdateSchemaIfNecessary is observed, the table has over 1 thousands columns. seems the loops on `tbl.Meta().Columns` caused high cpu usage, need to see if anyway to optimize the cpu usage when the number of column on the table is high.

image

image

```
			// Check the column change.
			for _, col := range tbl.Meta().Columns {
				if col.State != model.StatePublic {
					continue
				}
				found := false
				for _, coll := range tableInfo.Columns {
					if col.Name.L == coll.Name.L && col.ID != coll.ID {
						logutil.BgLogger().Info("public column changed", zap.String("column", col.Name.L), zap.String("old_col", coll.Name.L), zap.Int64("new id", col.ID), zap.Int64("old id", coll.ID))
						found = true
						break
					}
				}
				if found {
					if !skipLock {
						sctx.GetSessionVars().GetRelatedTableForMDL().Delete(tableInfo.ID)
					}
					return nil, domain.ErrInfoSchemaChanged.GenWithStack("public column %s has changed", col.Name)
				}
			}
		}
```

### PR DESCRIPTION
**Pull request by joechenrh**
### What problem does this PR solve?

Issue Number: close #54779

### What changed and how does it work?

Before checking table change by comparing `TableInfo`, we can just examine `Revision` first to skip unnecessary checks. Besides, I also found that column comparation is done by double loop, which can use `map`.

https://github.com/pingcap/tidb/blob/da7d83f54a8673016dd32415d5997f24f78c05ea/pkg/planner/core/preprocess.go#L1874-L1875

Using a table containing more than 1000 columns to test.

Before:

before

After:

Screenshot 2024-07-26 at 17 15 57

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/planner/core/preprocess.go [hunk 1]
```diff
@@ -1872,20 +1872,20 @@ func tryLockMDLAndUpdateSchemaIfNecessary(ctx context.Context, sctx base.PlanCon
 			sctx.GetSessionVars().GetRelatedTableForMDL().Store(tbl.Meta().ID, domainSchemaVer)
 		}
 		// Check the table change, if adding new public index or modify a column, we need to handle them.
-		if !sctx.GetSessionVars().IsPessimisticReadConsistency() {
+		if tbl.Meta().Revision != tableInfo.Revision && !sctx.GetSessionVars().IsPessimisticReadConsistency() {
 			var copyTableInfo *model.TableInfo
+
+			infoIndices := make(map[string]int64, len(tableInfo.Indices))
```
Comment (lcwangchao) [inline pkg/planner/core/preprocess.go:1878]: A indexID to indexName mapping with type `map[int64]int64` maybe more efficient in map searching

**Review by lcwangchao — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Comment by qw4990** [comment]
Do we need to back port this to v8.1 or v7.5?

**Pull request merged**